Step3 자판기 화면 구성

seizze opened the pull request:
### 고민한 점

* Notification Center에 옵저버 등록 시에 반복되는 코드가 많은 것 같아서 NotificationCenter에 extension을 추가하여 간소화하였습니다.
* IBOutletCollection을 사용할 때 뷰마다 tag를 지정해준 것을 `ViewIdentifier`에 모아서 관리하도록 만들었습니다. 처음엔 tag가 Int 타입이라서 Int에 extension을 추가할까 생각했었는데, 호출하는 쪽에서 `sender.tag.money`라고 하면 "태그에 있는 돈"으로 읽히는 것이 직관적이지 않은 것 같아서 `ViewIdentifier.findButton(by: sender.tag).money` 형식으로 접근하도록 만들었습니다.

seizze committed:
feat: Add beverage image assets

seizze committed:
feat: Setup vending machine storyboard UI

seizze committed:
feat: Add IBOutlet connections

seizze committed:
feat: Set tags to all the outlet collections

seizze committed:
feat: Update balance using notification

seizze committed:
feat: Change to identify which button is touched

seizze committed:
refactor: Separate notification name extension

seizze committed:
feat: Add beverage when button touched

seizze committed:
feat: Update labels when inventory changed

seizze committed:
refactor: Remove unused lines

seizze committed:
docs: Update README.md

godrm commented on VendingMachineApp/VendingMachineApp/Utils/ViewIdentifier.swift:
```diff
@@ -0,0 +1,58 @@
+            case .georgia: return "Georgia"
+            case .top: return "Top"
+            }
+        }
```
클래스로 선언한 세부적인 타입이 있는 데 매칭을 위해서 새로운 값을 정의한 부분은 아쉽습니다.
이미 선언한 타입을 활용하면 더 좋았을 것 같습니다.
Identifier나 ViewIdentifier가 화면 구성이 바뀔 때마다 코드가 바뀌어야 하는거라면 취약한 구조를 가지게 됩니다. 
이런 전담 객체를 만들려면 일반적인 역할을 하는 객체를 만들고 ViewController가 뷰 구조에 맞게 활용하는 어떨까요?

seizze replied:
@godrm 이것도 한참 고민해 봤는데 잘 모르겠습니다 😢...
재고가 몇개씩 있는지 알려주는 함수에서 `"\(type(of: beverage))"`를 딕셔너리의 키로 저장해버려서 재고를 알려면 `"Georgia"`, `"Top"`와 같은 키를 알아야할 것 같은데 키를 문자열로 하지 않도록 바꾸는게 좋을까요..?

godrm replied:
네. 타입에 대한 새로운 키를 만들면 클래스를 만들때마다 계속 변경해야 하는 상황이 생깁니다.
타입 자체를 활용할 수 있으면 가장 좋습니다. 
`type(of:)` 를 쓰고 있다면 음료 객체를 그대로 활용할 수 있지 않을까요?

seizze replied:
네! 일반적인 역할을 하는 UILabel의 서브클래스를 만들고 그 안에서 태그를 통해 무슨 레이블인지 판단하여 돌려주도록 수정하였습니다. 레이블의 텍스트를 업데이트하는 일도 이 서브클래스 안에서 하는것이 맞다고 생각이 들어서 같이 추가하였습니다.

인벤토리의 `briefStock()`에서도 음료 객체를 그대로 키로 갖도록 변경하였습니다.

카드게임 때도 비슷한 리뷰를 해주셨었는데 또 아무 생각 없이 이렇게 해버렸네요.ㅠㅠ 앞으로 타입을 선언할 때 모델과 뷰를 그저 매칭하기 위해 불필요한 타입을 만들고 있지 않은지 좀더 고민하겠습니다!

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -10,14 +10,82 @@ import UIKit
+    
+    @IBAction func addBeverage(_ sender: UIButton) {
+        if let id = ViewIdentifier.findBeverageButton(by: sender.tag) {
+            vendingMachine.fill(beverage: BeverageFactory.make(by: id))
```
ViewIdentifier와 BeverageFactory 관계를 잘 모르겠습니다. 
필요하다는 것도 알겠는데, 음료 메뉴 구성이나 음료 관계를 바꿀 때마다 저 객체를 바꿔야 한다는 부분이 조금 아쉽습니다.
물론 ViewController를 바꾸지 않아도 된다는 구조에서는 너무 좋습니다. 
UIButton.Tag를 쓰는게 아니라 Button을 세분화해보는 건 어떨까요? 
이미 쓰고 있는 OutletCollection 구조가 있으니 찾을 수 있을 것 같습니다. 
클래스를 사용하면 참조를 활용할 수 있다는 장점이 있습니다.

seizze replied:
Button을 세분화한다는것의 의미를 잘 모르겠습니다..
지금 생각나는 건..
```swift
struct Button {
    let button: UIButton // 버튼 참조?
    let beverageCreator: () -> Beverage // 어떤 음료수를 생성할지를 알고있는 클로저
}

class ViewController {
    var buttons = [Button]()
    func 버튼터치됨() {
        // 터치된버튼 beverageCreator() 동작시켜 벤딩머신에 전달
    }
    // ...
}
```
이런식으로 해서 초기화시에 클로저를 등록해놓고 활용하는것인데 이렇게 해도 될까요..?

godrm replied:
세분화하는 건 UIButton을 상속받아 새 클래스를 만드는거죠 :)

seizze replied:
이 부분도 BeverageButton이라는 UIButton 서브클래스를 만들어서 버튼에 대한 액션을 담당하도록 해봤습니다. 이 서브클래스 안에서 태그에 따라 어떤 음료수 인스턴스를 생성할지 분기 처리를 하였습니다. 생성된 인스턴스를 자판기에 추가하는 일은 컨트롤러에서 하는게 어울리는 것 같아서 ViewController에 작성하였습니다. 이와 같은 방법으로 금액 추가 버튼도 수정하였습니다!

음료 메뉴나 관계가 바뀔 때마다 변경되어야 하는 문제점이 있는 ViewIdentifier나 BeverageFactory는 개선 후에는 필요가 없어져서 둘다 삭제하였습니다.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -10,14 +10,82 @@ import UIKit
+            guard let cashier = $0 as? Cashier else { return }
+            self?.balanceLabel.text = "잔액: \(cashier.balance())원"
+        }
+        observers.append(observer)
```
좋은 시도였습니다. 이렇게 일반화할 꺼면 Observers 객체를 만드는 건 어떨까요? 
이름과 클로저만 전달하면 될 것 같네요. 
observers 배열도 여기에 직접 갖고 있지 않아도 되니까요.

seizze replied:
네 Observers 객체를 만들어 옵저버들을 관리하도록 하였습니다!

godrm reviewed (commented):
의미있는 시도를 한 것 같습니다. 그만큼 피드백도 많지 않습니다.
지금 구조에서 세부적으로 조금만 개선해보고 넘어갑시다. 다음 단계에도 지금 구조 덕분에 개선할 사항이 많지 않다면 좋은 설계를 한 게 확실할 겁니다.

seizze committed:
refactor: Abstract observer arrays to Obervers

seizze committed:
refactor: Move some functions to UIButton subclass

* Remove mediation type for distinguishing which button is touched
* Move initiating proper beverages to UIButton subclass

seizze committed:
refactor: Add balance button class

seizze committed:
refactor: Improve the label to know its identity

Improve the label to know its identity as a beverage type
Improve stock dictionary to have beverage object as keys
Remove ViewIdentifier type

seizze committed:
test: Fix inventory stock test to success

seizze committed:
docs: Update README.md

seizze commented:
피드백 모두 반영하였습니다 🙂

기능이 추가될 때 기존 코드가 변경되지 않는 유연한 구조를 설계하는 방법에 대해 전보다 훨씬 잘 이해하게 된 것 같습니다! 감사합니다 💪

godrm commented on VendingMachineApp/VendingMachineApp/Views/BeverageButton.swift:
```diff
@@ -0,0 +1,60 @@
+            case .georgia: return { return Georgia() }
+            case .top: return { return Top() }
+            }
+        }
```
이 부분도 버튼마다 다른 동작을 넣어놓고 비교 구문이 없었으면 더 좋았을 것 같네요. 음료수 배치나 종류가 달라졌을 때 바꾸지
않을 방법을 고민한다고 생각하면 힌트가 될 겁니다. 수고하셨어요

seizze replied:
앗 넵 더 고민해보겠습니다 😆
